pkg/tfbridge/x/token_test.go

```diff
@@ -472,46 +473,333 @@ func TestAliasing(t *testing.T) {
+                        "maxItemOne": false
+                    },
+                    "f2": {
+                        "maxItemOne": false,
+                        "fields": {
+                            "n1": {
+                                "maxItemOne": true
+                            },
+                            "n2": {
+                                "maxItemOne": false
+                            }
+                        }
+                    }
+                }
+            }
+        },
+        "datasources": {}
+    }
+}`
+	assert.Equal(t, expected, v)
+
+	// Apply the saved metadata to a new provider
+	info = provider(false, true)
+	err = AutoAliasing(info, metadata)
+	require.NoError(t, err)
+
+	assert.Equal(t, expected, string(metadata.Marshal()))
+	assert.True(t, *info.Resources["pkg_r1"].Fields["f2"].Fields["n1"].MaxItemsOne)
+	assert.False(t, *info.Resources["pkg_r1"].Fields["f2"].Fields["n2"].MaxItemsOne)
+}
+
 type Provider struct {
 	util.UnimplementedProvider
 
 	// We are only concerned with tokens, so that's all we support
-	datasources map[string]struct{}
-	resources   map[string]struct{}
+	datasources map[string]shim.SchemaMap
+	resources   map[string]shim.SchemaMap
 }
 
-func (p Provider) ResourcesMap() shim.ResourceMap   { return ResourceMap{p.resources} }
-func (p Provider) DataSourcesMap() shim.ResourceMap { return ResourceMap{p.datasources} }
+func (p Provider) ResourcesMap() shim.ResourceMap   { return newResourceMap(p.resources) }
+func (p Provider) DataSourcesMap() shim.ResourceMap { return newResourceMap(p.datasources) }
 
-type ResourceMap struct{ m map[string]struct{} }
-type Resource struct{ t string }
+type Schema struct {
+	shim.Schema
+	MaxItemsOne bool
+	elem        any
+}
 
-func (m ResourceMap) Len() int                     { return len(m.m) }
-func (m ResourceMap) Get(key string) shim.Resource { return Resource{key} }
-func (m ResourceMap) GetOk(key string) (shim.Resource, bool) {
-	_, ok := m.m[key]
-	if !ok {
-		return nil, false
+func (s Schema) MaxItems() int {
+	if s.MaxItemsOne {
+		return 1
 	}
-	return Resource{key}, true
+	return 0
+}
+
+func (s Schema) Type() shim.ValueType {
+	return shim.TypeList
 }
-func (m ResourceMap) Range(each func(key string, value shim.Resource) bool) {
-	for k := range m.m {
-		each(k, Resource{k})
+
+func (s Schema) Elem() any { return s.elem }
+
+// type ResourceMap struct{ m map[string]shim.SchemaMap }
+type Resource struct {
+	shim.Resource
+	t      string
+	schema shim.SchemaMap
+}
+
+func newResourceMap(m map[string]shim.SchemaMap) shim.ResourceMap {
+	return &mapKind[shim.Resource, shim.SchemaMap]{
+		m: m,
+		getTransform: func(k string, v shim.SchemaMap) shim.Resource {
+			return Resource{nil, k, v}
+		},
+		setTransform: func(r shim.Resource) shim.SchemaMap {
+			return r.Schema()
+		},
+	}
+}
+
+func (r Resource) Schema() shim.SchemaMap {
+	return r.schema
+}
+
+func newSchemaMap(m map[string]shim.Schema) shim.SchemaMap {
```

Comment:
FWIW there's some existing structs implementing shim.SchemaMap in the shim module, not a blocker..